src/api/procedures/addConfidentialTransaction.ts

```diff
@@ -0,0 +1,317 @@
+      !legLengthErrIndexes.length &&
+      !legAmountErrIndexes.length &&
+      !sameIdentityErrIndexes.length
+    ) {
+      const rawVenueId = bigNumberToU64(venueId, context);
+      const rawLegs: PalletConfidentialAssetTransactionLeg[] = [];
+      const rawInstructionMemo = optionize(stringToMemo)(memo, context);
+
+      await Promise.all([
+        ...legs.map(
+          async ({
+            sender: inputSender,
+            receiver: inputReceiver,
+            assets: inputAssets,
+            auditors: inputAuditors,
+            mediators: inputMediators,
+          }) => {
+            const assets = inputAssets.map(asset => asConfidentialAsset(asset, context));
+            const sender = asConfidentialAccount(inputSender, context);
+            const receiver = asConfidentialAccount(inputReceiver, context);
+            const auditors = inputAuditors.map(auditor => asConfidentialAccount(auditor, context));
+            const mediators = inputMediators.map(mediator => asIdentity(mediator, context));
+            await Promise.all([
+              assertConfidentialAccountExists(sender),
+              assertConfidentialAccountExists(receiver),
+              ...assets.map(asset => assertConfidentialAssetExists(asset, context)),
+              ...auditors.map(auditor => assertConfidentialAccountExists(auditor)),
+              ...mediators.map(mediator => assertIdentityExists(mediator)),
+            ]);
+
+            const rawSender = confidentialAccountToMeshPublicKey(sender, context);
+            const rawReceiver = confidentialAccountToMeshPublicKey(receiver, context);
+            const rawAuditors = auditorsToBtreeSet(auditors, context);
+            const rawMediators = identitiesToBtreeSet(mediators, context);
+            const rawAssets = confidentialAssetsToBtreeSet(assets, context);
+
+            const rawLeg = confidentialLegToMeshLeg(
+              {
+                sender: rawSender,
+                receiver: rawReceiver,
+                auditors: rawAuditors,
+                mediators: rawMediators,
+                assets: rawAssets,
+              },
+              context
+            );
+
+            rawLegs.push(rawLeg);
+          }
+        ),
+      ]);
+
+      addTransactionParams.push([rawVenueId, rawLegs, rawInstructionMemo]);
+    }
+  }
+
+  return {
+    errIndexes: {
+      legEmptyErrIndexes,
+      legLengthErrIndexes,
+      sameIdentityErrIndexes,
+      noAssetsErrIndexes,
+    },
+    addTransactionParams,
+  };
+}
+
+/**
+ * @hidden
+ */
+export async function prepareAddTransaction(
+  this: Procedure<Params, ConfidentialTransaction[]>,
+  args: Params
+): Promise<BatchTransactionSpec<ConfidentialTransaction[], unknown[][]>> {
+  const {
+    context: {
+      polymeshApi: {
+        tx: { confidentialAsset },
+      },
+    },
+    context,
+  } = this;
+  const { transactions, venueId } = args;
+
+  await assertConfidentialVenueExists(venueId, context);
+
```

Comment:
We also need to add check if venue filtering allows that particular venue to add transaction for a given asset id

Reply:
added a check on line 164